File: webui/staging.py
"""Stage selected event minutes for the engine.

Jobs get a staging folder of symlinks so the engine only sees the selected
one-minute clips — non-contiguous selections work without any engine changes.
Previews get short stream-copied trims instead so the engine encodes only a
couple of seconds.
"""

from __future__ import annotations

import logging
import shutil
import subprocess
from pathlib import Path
from typing import Any, Optional

from . import config
from .scanner import CLIP_RE, minute_files

_LOGGER = logging.getLogger(__name__)


def _fresh_dir(path: Path) -> Path:
    if path.exists():
        shutil.rmtree(path)
    path.mkdir(parents=True)
    return path


def stage_job(job_id: str, selections: list[dict[str, Any]]) -> Path:
    """Create a staging root with one folder per selected event.

    Each ``selection`` is ``{"dir": Path, "minutes": [keys] | None}`` and gets
    a ``staged_name`` key set on it (unique folder name inside the staging
    root, which also becomes the output movie name).
    """
    root = _fresh_dir(config.STAGING_DIR / job_id)
    used_names: set[str] = set()

    for selection in selections:
        source_dir: Path = selection["dir"]
        name = source_dir.name
        if name in used_names:
            suffix = 2
            while f"{name}_{suffix}" in used_names:
                suffix += 1
            name = f"{name}_{suffix}"
        used_names.add(name)
        selection["staged_name"] = name

        staged_dir = root / name
        staged_dir.mkdir()

        files = minute_files(source_dir, selection.get("minutes") or None)
        if not files:
            raise ValueError(f"no clips selected in event folder {source_dir.name!r}")
        for file in files:
            (staged_dir / file.name).symlink_to(file.resolve())

        event_json = source_dir / "event.json"
        if event_json.is_file():
            shutil.copy2(event_json, staged_dir / "event.json")

    return root


def stage_preview(
    preview_id: str,
    event_dir: Path,
    minute_key: str,
    cameras: Optional[list[str]] = None,
    seconds: int = 3,
) -> Path:
    """Stage one minute as short stream-copied trims for a fast preview."""
    root = _fresh_dir(config.STAGING_DIR / f"preview-{preview_id}")

    files = minute_files(event_dir, [minute_key], cameras)
    if not files:
        raise ValueError(f"no clips found for minute {minute_key!r}")

    for file in files:
        subprocess.run(
            [
                config.FFMPEG,
                "-y",
                "-loglevel",
                "error",
                "-i",
                str(file),
                "-t",
                str(seconds),
                "-c",
                "copy",
                str(root / file.name),
            ],
            check=True,
            capture_output=True,
            timeout=60,
        )

    event_json = event_dir / "event.json"
    if event_json.is_file():
        shutil.copy2(event_json, root / "event.json")
    return root


def cleanup(path: Path) -> None:
    """Remove a staging folder, ignoring errors."""
    if path and path.exists():
        shutil.rmtree(path, ignore_errors=True)


def directory_size(path: Path) -> int:
    """Total size in bytes of all real files under path (symlinks excluded)."""
    total = 0
    for entry in path.rglob("*"):
        try:
            if entry.is_file() and not entry.is_symlink():
                total += entry.stat().st_size
        except OSError:
            continue
    return total

File: macos/DashcamStudio/Staging.swift
//  Staging.swift  (WP-2)
//  Stage selected minutes as a temp folder of symlinks so the engine only
//  sees the chosen clips — non-contiguous selection works with zero engine
//  changes. Port of the tested webui staging.py. Reads in place (symlinks).

import Foundation

struct Staging: Sendable {
    let scanner = TeslaCamScanner()
    private var fm: FileManager { .default }

    struct StagedEvent {
        let stagedName: String
        let stagedDir: URL
        let sourceDir: URL
    }

    /// Create a fresh staging root with one symlinked folder per event.
    /// `selections` maps a source event folder → chosen minute keys (nil = all).
    func stage(jobID: JobID,
               cacheRoot: URL,
               selections: [(source: URL, minutes: Set<String>?)]) throws -> (root: URL, events: [StagedEvent]) {
        let root = cacheRoot.appending(path: "staging/\(jobID.uuidString)")
        try? fm.removeItem(at: root)
        try fm.createDirectory(at: root, withIntermediateDirectories: true)

        var used = Set<String>()
        var staged: [StagedEvent] = []
        for sel in selections {
            var name = sel.source.lastPathComponent
            if used.contains(name) {
                var i = 2
                while used.contains("\(name)_\(i)") { i += 1 }
                name = "\(name)_\(i)"
            }
            used.insert(name)

            let dir = root.appending(path: name)
            try fm.createDirectory(at: dir, withIntermediateDirectories: true)

            let files = scanner.clipURLs(in: sel.source, minutes: sel.minutes)
            guard !files.isEmpty else { throw EngineError.noClips }
            for f in files {
                try fm.createSymbolicLink(at: dir.appending(path: f.lastPathComponent),
                                          withDestinationURL: f.resolvingSymlinksInPath())
            }
            let eventJSON = sel.source.appending(path: "event.json")
            if fm.fileExists(atPath: eventJSON.path) {
                try? fm.copyItem(at: eventJSON, to: dir.appending(path: "event.json"))
            }
            staged.append(StagedEvent(stagedName: name, stagedDir: dir, sourceDir: sel.source))
        }
        return (root, staged)
    }

    func cleanup(_ root: URL?) {
        guard let root else { return }
        try? fm.removeItem(at: root)
    }

    /// Total size of real files under a folder (symlink targets excluded).
    func directorySize(_ url: URL) -> Int64 {
        guard let en = fm.enumerator(at: url, includingPropertiesForKeys: [.isSymbolicLinkKey, .fileSizeKey]) else { return 0 }
        var total: Int64 = 0
        for case let f as URL in en {
            let v = try? f.resourceValues(forKeys: [.isSymbolicLinkKey, .fileSizeKey])
            if v?.isSymbolicLink == true { continue }
            total += Int64(v?.fileSize ?? 0)
        }
        return total
    }
}
